[CAM-13974] Preparing External Task Action Listeners

tahaarian opened the pull request:
Hello Team

I hope your are doing well.

According to our [negotiation](https://jira.camunda.com/browse/CAM-13974) on JIRA, I have sent a pull request for the point.

I have attached related story as below

**User Story (Required on creation):**

Following [this](https://forum.camunda.org/t/external-task-action-listeners/7431) post, preparing related actions (Listeners) will be important for the end-user client that need to work with Rest API (between External task and Engine). Actually, there are more than 10 Rest API that will call around the application and could be trigger related event that would be useful.

**Functional Requirements (Required before implementation):**

The main goal is to prepare module's listener for catching state of each. it should be corporate with an External Task and manage three state (on,done,fail).

As we describe the point, it will be like a package for listeners and make the main interface.

Thank you so much for your support and great feedbacks.

I will appreciate if you could check the request and let me know if you have any points.

Best Regards,

Taha Arian

P.S : please assign to Mr Tobias Metzke-Bernstein

tahaarian commented on clients/java/client/src/it/java/org/camunda/bpm/client/client/ClientIT.java:
```diff
@@ -252,6 +250,92 @@ public void shouldThrowExceptionDueToMaxTasksNotGreaterThanZero() {
+        } catch (InterruptedException ignored) {
+        }
+
+      List<String> firstStates = new ArrayList<>(assertStates.length);
+
+      for (int i = 0; i < assertStates.length; i++) firstStates.add(states.get(i));
```
hello again @tmetzke

I have copied it just for better assertion as `assertArrayEquals`  :slightly_smiling_face:  , because we have a `LinkedList (states)` that could extended according to our used listeners.

I think there is not any assertion concept for LinkedList, Let me know if you have any better suggestion :slightly_smiling_face: 

Thank you in advance

tmetzke replied:
I would encourage you to have a look at AssertJ instead of JUnit's Assert class. This is what we use in this class here anyways and can also do something like `assertThat(states).containsExactlyInAnyOrderElementsOf(assertStates)`. Play around with that and see if something matches what you would like to assert here.

tahaarian commented on clients/java/client/src/it/java/org/camunda/bpm/client/client/ClientIT.java:
```diff
@@ -252,6 +250,92 @@ public void shouldThrowExceptionDueToMaxTasksNotGreaterThanZero() {
+        try {
+          Thread.sleep(10);
+        } catch (InterruptedException ignored) {
+        }
```
Thank you for your great response. I could understand what you said totally but can not find which example you mean. 

I have changed this part to a new checker. Let me know if you prefer other style :slightly_smiling_face:

tahaarian commented:
Hello again @tmetzke

Your welcome my friend :slightly_smiling_face: 
Thank you in advance for your great response and nice feedbacks. I have read related document(again) as you suggest and make some changes according to your points. 

1- all your suggestions were great and accepted :+1: 
2- create full override for rest of methods
3- prepare better `assertStates` (let me know if we need more function for test or etc)
4- change infinite loop to `ClientRule` type (let me know if its ok or not, i can not find better solution)

In the end, I have wrote some comments on your review parts too. 

Thank you for your support 

Best,

Taha Arian

tahaarian commented on clients/java/client/src/main/java/org/camunda/bpm/client/impl/RequestExecutor.java:
```diff
@@ -37,6 +37,7 @@
 import org.apache.http.message.BasicHeader;
 import org.apache.http.util.EntityUtils;
 import org.camunda.bpm.client.interceptor.impl.RequestInterceptorHandler;
+import org.camunda.bpm.client.listener.ClientInteractionListener;
```
yes

tahaarian commented on clients/java/client/src/main/java/org/camunda/bpm/client/impl/RequestExecutor.java:
```diff
@@ -57,6 +58,8 @@
   protected HttpClient httpClient;
   protected ObjectMapper objectMapper;
 
+  protected ClientInteractionListener clientInteractionListener;
+
```
yes

tahaarian commented on clients/java/client/src/main/java/org/camunda/bpm/client/interceptor/impl/RequestInterceptorHandler.java:
```diff
@@ -24,6 +24,7 @@
 import org.camunda.bpm.client.impl.EngineClientLogger;
 import org.camunda.bpm.client.impl.ExternalTaskClientLogger;
 import org.camunda.bpm.client.interceptor.ClientRequestInterceptor;
+import org.camunda.bpm.client.listener.ClientInteractionListener;
```
yes

tmetzke commented on clients/java/client/src/main/java/org/camunda/bpm/client/interceptor/impl/RequestInterceptorHandler.java:
```diff
@@ -24,7 +24,6 @@
 import org.camunda.bpm.client.impl.EngineClientLogger;
 import org.camunda.bpm.client.impl.ExternalTaskClientLogger;
 import org.camunda.bpm.client.interceptor.ClientRequestInterceptor;
-
```
I think you can revert all changes of this class, so the commit for the PR is clean in the end.

tahaarian replied:
Yes, you are right. reverted 
thanks

tmetzke commented on clients/java/client/src/main/java/org/camunda/bpm/client/impl/RequestExecutor.java:
```diff
@@ -57,6 +57,7 @@
   protected HttpClient httpClient;
   protected ObjectMapper objectMapper;
 
+
```
I think you can revert all changes of this class, so the commit for the PR is clean in the end.

tahaarian replied:
Yes, you are right. reverted 
thanks

tmetzke commented on clients/java/client/src/it/java/org/camunda/bpm/client/client/ClientIT.java:
```diff
@@ -39,19 +43,13 @@
 import java.io.FileNotFoundException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Date;
-import java.util.List;
-import java.util.Properties;
+import java.util.*;
```
We try to avoid star imports, as they might create conflicts

tahaarian replied:
good point, its done :slightly_smiling_face:

tmetzke reviewed (changes requested):
Hi @tahaarian,

that looks pretty good already, well done!

I have some minor last points:

* Let's try to avoid star imports. They can create conflicts if methods exist in multiple namespaces.
* Try to have a look at AssertJ for the test class.
* Indentation still looks a bit off in the test class, all other classes look fine in that regard.
* I think renaming the `externalTaskClientListener` method to `clientInteractionListener` better reflects what it does.

Best,
Tobias

tahaarian commented:
Hello again @tmetzke

Thank you for your quick response and great suggestions.

I have modified all points that you suggested and make new assertion according to `AssertJ`

I will appreciated if you could check them out.

Let me know if you have new feedbacks.

Best,
Taha Arian

tmetzke commented:
Hi @tahaarian,

there is one additional issue we need to address in this PR: exception classes.

Some of the interface methods in `ClientInteractionListener` consume an `EngineClientException` as a parameter. This class is however private API (you can see that its package name contains an `impl`) and therefore not suited for a public interface. I would suggest the following here:

* Replace the `EngineClientException` with `ExternalTaskClientException` in `ClientInteractionListener` and `DefaultClientInteractionListener`
* In the `ExternalTaskServiceImpl`, pass on the exceptions created by `LOG#externalTaskServiceException` instead of the internal `EngineClientException`.
* For the exception used in the fetch and lock call in the `TopicSubscriptionManager`:
  * Merge the current master branch into your PR (the fetch and lock has been adjusted slightly).
  * Reuse the exception created by `LOG#fetchAndLockException`.

Let me know if you have any further questions.

Best,
Tobias

tahaarian commented:
Hello again @tmetzke
Thanks for your great feedback and sorry for delay.

I have modified related part according to your suggestions. Let me explain some part :

1- `RequestInterceptorHandler.java` and `RequestExecutor.java` were reverted.

2- I have created new test class (`ClientInteractionListenerIT`) and improve testing scenario according to your suggestion. Then i have checked all possible(tests) interactions with the `External Task API`  (all listeners were covered successfully).

3- I have reverted `ClientIT `to the orginal state.

4- I have merged `TopicSubscriptionManager` state with our modification according to your point.

5- I think about exceptions part and make a new modification. Actually you handled all type of exceptions on your `externalTaskServiceException` method and its a great exception manager. So i prefer to delete exception instance on our `OnFail` handler (for example `setVariableFail `and etc) because it was redundant usage. I think it could be more simple implementation.

I will appreciate if you could check them out. Let me know your feedbacks.

Thank you in advance

Taha Arian

tmetzke commented:
Hi @tahaarian,

thanks for your input on this.
I will have a look at it next week, sorry for the delay.

Best,
Tobias

tahaarian commented:
Hello again @tmetzke

Your welcome. 
Thanks to you for your great support 

Taha Arian

tmetzke reviewed (changes requested):
Hi @tahaarian,

thanks for your input and sorry for the delay.
Thanks for also extracting the test class for the interaction listener.

Some room for improvement I still see here:

* The classes `ClientIT`, `RequestExecutor`, and `RequestInterceptorHandler are not really reverted in the sense of git. There are still changes in those classes, you can see that in the **Files changed** tab of the PR. Please use `git checkout` to check out those classes with their original content and commit those again to this PR.
* I would actually rather keep the exceptions in the "failXYZ" API methods of the listener. This enables the consumer to also inspect the error when receiving the event. If someone would like to have that functionality later on, we would have to adjust the API. I don't see why we should not add this now.
* Sorry for asking again, but have you also run the tests locally? They do not succeed for me locally, especially the `ClientIT` fails due to a `NullPointerException`. I believe the client interaction listener is null per default in the `TopicSubscriptionManager`. However, the listener is never null-checked but always accessed as if it can never be null. Please make sure to run the tests locally before submitting here, as this prolongs the review immensely. You can run `mvn clean verify` on the client project locally to run all tests for example.
* Furthermore, the `ClientInteractionListenerIT` is failing as well, as not all states have been reached as expected. I believe that you need to run multiple test cases with different models in order to pass by all expected listener methods in the end. It might make sense to initialize the client in the test class with the same interaction listener for all test methods then, so you don't have to create it anew all the time. Besides, I think it might be also more flexible to use a mock class for the interaction listener and use [Mockito's verify mechanism](https://www.baeldung.com/mockito-verify) to assert that the API methods have been called. No need to collect states, I think.

Let me know if anything is unclear.

Best,
Tobias

tmetzke commented:
Hi @tahaarian,

this is a quick check-in to see if you're still interested in moving forward with this. 🙂 

Best,
Tobias

tahaarian commented:
Hi @tmetzke 

Thanks for your feedback. Yes, I am working on it (but a little busy with parallel tasks). 
I will notify you asap :slightly_smiling_face: 

Best Regards,
Taha Arian

tmetzke commented:
Hi @tahaarian,

thanks for the input. No worries, take your time.

Best,
Tobias

tmetzke commented:
Hi @tahaarian,

I wanted to check back in with you. This has been stale for a while now.
Should we keep this open or close it for now?

In any case, you could simply reopen it at your leisure in case you find the time to work on it.

Best,
Tobias

tmetzke commented:
Hi @tahaarian,

I will close this for now due to inactivity. Feel free to reopen any time at your leisure.

Best,
Tobias

tahaarian commented:
Hello @tmetzke 

Sry for my delay, we were busy for some tasks. 

Thank you, so I will reopen it whenever i need to work on 

Best Regards,
Taha
